numpyro/distributions/copula.py

```diff
@@ -0,0 +1,151 @@
+# Copyright Contributors to the Pyro project.
+# SPDX-License-Identifier: Apache-2.0
+
+from jax import lax, numpy as jnp
+
+import numpyro.distributions.constraints as constraints
+from numpyro.distributions.continuous import Beta, MultivariateNormal, Normal
+from numpyro.distributions.distribution import Distribution
+from numpyro.distributions.util import (
+    is_prng_key,
+    lazy_property,
+    promote_shapes,
+    validate_sample,
+)
+
+
+class GaussianCopula(Distribution):
+    """
+    A distribution that links the batch_shape[:-1] of marginal distribution `marginal_dist`
+    with a multivariate Gaussian copula modelling the correlation between the axes.
+
+    :param Distribution marginal_dist: Distribution whose last batch axis is to be coupled.
+    :param array_like correlation_matrix: Correlation matrix of coupling multivariate normal distribution.
+    :param array_like correlation_cholesky: Correlation Cholesky factor of coupling multivariate normal distribution.
+    """
+
+    arg_constraints = {
+        "correlation_matrix": constraints.corr_matrix,
+        "correlation_cholesky": constraints.corr_cholesky,
+    }
+    reparametrized_params = [
+        "correlation_matrix",
+        "correlation_cholesky",
+    ]
+
+    def __init__(
+        self,
+        marginal_dist,
+        correlation_matrix=None,
+        correlation_cholesky=None,
+        *,
+        validate_args=None,
+    ):
+        if len(marginal_dist.event_shape) > 0:
+            raise ValueError("`marginal_dist` needs to be a univariate distribution.")
+
+        self.marginal_dist = marginal_dist
+        self.base_dist = MultivariateNormal(
+            covariance_matrix=correlation_matrix,
+            scale_tril=correlation_cholesky,
+        )
+
+        event_shape = self.base_dist.event_shape
+        batch_shape = lax.broadcast_shapes(
+            self.marginal_dist.batch_shape[:-1],
+            self.base_dist.batch_shape,
+        )
+
+        super(GaussianCopula, self).__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+    def sample(self, key, sample_shape=()):
+        assert is_prng_key(key)
+
+        shape = sample_shape + self.batch_shape
+        normal_samples = self.base_dist.expand(shape).sample(key)
+        cdf = Normal().cdf(normal_samples)
+        return self.marginal_dist.icdf(cdf)
+
+    @validate_sample
+    def log_prob(self, value):
+        # Ref: https://en.wikipedia.org/wiki/Copula_(probability_theory)#Gaussian_copula
+        # see also https://github.com/pyro-ppl/numpyro/pull/1506#discussion_r1037525015
+        marginal_lps = self.marginal_dist.log_prob(value)
+        quantiles = Normal().icdf(self.marginal_dist.cdf(value))
```

Comment:
@hessammehr I guess you can clip here. cdf might return 0 or 1 so we can clip by tiny and eps. You can use `util.clamp_probs` for this